balboa-client-jms/src/test/scala/com/blist/metrics/impl/queue/MetricJmsQueueSpec.scala

```diff
@@ -0,0 +1,177 @@
+      MetricRecord(1496268557542l, "14", "num-metadata-available", 270,  Metric.RecordType.ABSOLUTE)
+    )
+  }
+
+  trait AllAggregateQueueSetup extends QueueSetup {
+    val testMetrics: IndexedSeq[MetricRecord] = IndexedSeq(
+      MetricRecord(1496268557542l, "3", "num-metadata-total", 21,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "3", "num-metadata-total", 114,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "1", "num-metadata-total", 5339,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "1", "num-metadata-total", 20916,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "14", "num-metadata-total", 41,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "14", "num-metadata-total", 270,  Metric.RecordType.ABSOLUTE)
+    )
+  }
+
+  trait MixedQueueSetup extends QueueSetup {
+    val testMetrics: IndexedSeq[MetricRecord] = IndexedSeq(
+      MetricRecord(1496268557542l, "3", "num-metadata-completed", 21,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "3", "num-metadata-available", 114,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "3", "num-metadata-total", 21,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "3", "num-metadata-total", 114,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "1", "num-metadata-completed", 5339,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "1", "num-metadata-available", 20916,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "1", "num-metadata-total", 5339,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "1", "num-metadata-total", 20916,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "14", "num-metadata-completed", 41,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "14", "num-metadata-available", 270,  Metric.RecordType.ABSOLUTE),
+      MetricRecord(1496268557542l, "14", "num-metadata-total", 41,  Metric.RecordType.AGGREGATE),
+      MetricRecord(1496268557542l, "14", "num-metadata-total", 270,  Metric.RecordType.AGGREGATE)
+    )
+  }
+
+  abstract trait QueueSetup {
+    val mockQueue: Queue = mock[Queue]
+
+    val mockProducer: MessageProducer = mock[MessageProducer]
+
+    val mockSession: Session = mock[Session]
+
+    when(mockSession.createQueue("not-a-real-queue")).thenReturn(mockQueue)
+    when(mockSession.createProducer(mockQueue)).thenReturn(mockProducer)
+
+    val mockConnection: Connection = mock[Connection]
+
+    when(mockConnection.createSession(false, Session.AUTO_ACKNOWLEDGE)).thenReturn(mockSession)
+
+    case class MetricRecord(timestamp: Long, entityId: String, name: String, value: Number, recordType: Metric.RecordType)
+
+    def testMetrics: IndexedSeq[MetricRecord]
+
+    val metricWithZeroValue = new Metric()
+    metricWithZeroValue.setValue(0)
+    metricWithZeroValue.setType(Metric.RecordType.AGGREGATE)
+
+    def run() = {
```

Comment:
This seems pretty complicated logic for a test, but maybe there's no other way...